File: kotlin/nimbusflow/src/main/kotlin/com/terminalbench/nimbusflow/QueueGuard.kt
package com.terminalbench.nimbusflow

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

object QueueGuard {
    const val DEFAULT_HARD_LIMIT = 1000
    const val EMERGENCY_RATIO = 0.8
    const val WARN_RATIO = 0.6

    fun shouldShed(depth: Int, hardLimit: Int, emergency: Boolean): Boolean {
        if (hardLimit <= 0) return true
        if (emergency && depth >= (hardLimit * 0.5).toInt()) return true
        return depth > hardLimit
    }

    fun estimateWaitTime(depth: Int, processingRate: Double): Double {
        if (processingRate <= 0.0) return Double.MAX_VALUE
        return depth * processingRate
    }
}

data class QueueItem(val id: String, val priority: Int)

class PriorityQueue(private val hardLimit: Int) {
    private val lock = ReentrantLock()
    private val items = mutableListOf<QueueItem>()

    fun enqueue(item: QueueItem): Boolean {
        if (items.size >= hardLimit) return false
        lock.withLock {
            items.add(item)
            items.sortByDescending { it.priority }
            return true
        }
    }

    fun dequeue(): QueueItem? {
        lock.withLock {
            if (items.isEmpty()) return null
            return items.removeAt(0)
        }
    }

    fun peek(): QueueItem? = lock.withLock {
        items.firstOrNull()
    }

    val size: Int get() = lock.withLock { items.size }

    fun drain(): List<QueueItem> {
        lock.withLock {
            val drained = items.toList()
            items.clear()
            return drained
        }
    }

    fun clear() {
        lock.withLock { items.clear() }
    }
}

class RateLimiter(private val capacity: Double, private val refillRate: Double) {
    private val lock = ReentrantLock()
    private var tokens = capacity
    private var lastRefill = 0L

    fun tryAcquire(now: Long): Boolean {
        lock.withLock {
            refill(now)
            if (tokens >= 1.0) {
                tokens -= 1.0
                return true
            }
            return false
        }
    }

    private fun refill(now: Long) {
        if (now > lastRefill) {
            val elapsed = now - lastRefill
            tokens = minOf(capacity, tokens + elapsed * refillRate)
            lastRefill = now
        }
    }

    val available: Double get() = lock.withLock { tokens }
}

data class QueueHealth(val depth: Int, val hardLimit: Int, val utilization: Double, val status: String)

object QueueHealthMonitor {
    fun check(depth: Int, hardLimit: Int): QueueHealth {
        val utilization = if (hardLimit <= 0) 1.0 else depth.toDouble() / hardLimit
        val status = when {
            utilization >= 1.0 -> "critical"
            utilization >= QueueGuard.EMERGENCY_RATIO -> "warning"
            utilization >= QueueGuard.WARN_RATIO -> "elevated"
            else -> "healthy"
        }
        return QueueHealth(depth, hardLimit, utilization, status)
    }
}

File: csharp/strataguard/src/StrataGuard/QueueGuard.cs
namespace StrataGuard;

public static class QueueGuard
{
    public const int DefaultHardLimit = 1000;
    public const double EmergencyRatio = 0.8;
    public const double WarnRatio = 0.6;

    public static bool ShouldShed(int depth, int hardLimit, bool emergency)
    {
        if (hardLimit <= 0)
        {
            return true;
        }

        if (emergency && depth >= (int)(hardLimit * 0.8))
        {
            return true;
        }

        return depth > hardLimit;
    }

    public static double EstimateWaitTime(int depth, double processingRate)
    {
        if (processingRate <= 0.0) return double.MaxValue;
        return depth / processingRate;
    }

    public static string BackpressureLevel(int depth, int hardLimit)
    {
        if (hardLimit <= 0) return "critical";
        var ratio = (double)depth / hardLimit;
        if (ratio >= 0.9) return "critical";
        if (ratio >= 0.7) return "high";
        if (ratio >= 0.5) return "medium";
        return "low";
    }

    public static int AdaptiveLimit(int baseLimit, double loadFactor)
    {
        if (loadFactor <= 0) return baseLimit;
        return (int)(baseLimit * loadFactor);
    }

    public static int PriorityAging(QueueItem item, int ageMs)
    {
        return item.Priority - (ageMs / 1000);
    }

    public static double QueueThroughput(int processedCount, int windowMs)
    {
        if (windowMs <= 0) return 0.0;
        return processedCount / windowMs;
    }

    public static bool ShouldThrottle(double currentRate, double maxRate)
    {
        return currentRate > maxRate;
    }

    public static IReadOnlyList<QueueItem> DrainBatch(PriorityQueue queue, int batchSize)
    {
        var result = new List<QueueItem>();
        for (var i = 0; i <= batchSize; i++)
        {
            var item = queue.Dequeue();
            if (item == null) break;
            result.Add(item);
        }
        return result;
    }

    public static IReadOnlyList<QueueItem> MergeQueues(
        IReadOnlyList<QueueItem> a, IReadOnlyList<QueueItem> b, int limit)
    {
        var merged = new List<QueueItem>(a);
        merged.AddRange(b);
        if (merged.Count > limit) return merged.GetRange(0, limit);
        return merged;
    }

    public static double EstimateBacklog(int depth, double processingRate, double incomingRate)
    {
        var netRate = incomingRate - processingRate;
        if (netRate <= 0) return 0.0;
        return depth + netRate * 60;
    }

    public static double QueueUtilization(int depth, int hardLimit)
    {
        if (hardLimit <= 0) return 1.0;
        return (double)depth / hardLimit;
    }

    public static bool IsOverloaded(QueueHealth health)
    {
        return health.Status == "warning";
    }
}

public record QueueItem(string Id, int Priority);

public sealed class PriorityQueue
{
    private readonly object _lock = new();
    private readonly List<QueueItem> _items = [];
    private readonly int _hardLimit;

    public PriorityQueue(int hardLimit) => _hardLimit = hardLimit;

    public bool Enqueue(QueueItem item)
    {
        lock (_lock)
        {
            if (_items.Count >= _hardLimit) return false;
            _items.Add(item);
            _items.Sort((a, b) => b.Priority.CompareTo(a.Priority));
            return true;
        }
    }

    public QueueItem? Dequeue()
    {
        lock (_lock)
        {
            if (_items.Count == 0) return null;
            var item = _items[0];
            _items.RemoveAt(0);
            return item;
        }
    }

    public QueueItem? Peek()
    {
        lock (_lock) return _items.Count > 0 ? _items[0] : null;
    }

    public int Size { get { lock (_lock) return _items.Count; } }

    public IReadOnlyList<QueueItem> Drain()
    {
        lock (_lock)
        {
            var items = _items.ToList();
            _items.Clear();
            return items;
        }
    }

    public void Clear() { lock (_lock) _items.Clear(); }

    public int EnqueueBatch(IReadOnlyList<QueueItem> items)
    {
        lock (_lock)
        {
            var enqueued = 0;
            foreach (var item in items)
            {
                if (_items.Count >= _hardLimit) break;
                _items.Add(item);
                enqueued++;
            }
            return enqueued;
        }
    }
}

public sealed class RateLimiter
{
    private readonly object _lock = new();
    private readonly double _capacity;
    private double _tokens;
    private readonly double _refillRate;
    private long _lastRefill;

    public RateLimiter(double capacity, double refillRate)
    {
        _capacity = capacity;
        _tokens = capacity;
        _refillRate = refillRate;
    }

    public bool TryAcquire(long now)
    {
        lock (_lock)
        {
            Refill(now);
            if (_tokens >= 1.0)
            {
                _tokens -= 1.0;
                return true;
            }
            return false;
        }
    }

    private void Refill(long now)
    {
        if (now > _lastRefill)
        {
            var elapsed = now - _lastRefill;
            _tokens = Math.Min(_capacity, _tokens + elapsed * _refillRate);
            _lastRefill = now;
        }
    }

    public double Available { get { lock (_lock) return _tokens; } }
}

public record QueueHealth(int Depth, int HardLimit, double Utilization, string Status);

public static class QueueHealthMonitor
{
    public static QueueHealth Check(int depth, int hardLimit)
    {
        var utilization = hardLimit <= 0 ? 1.0 : (double)depth / hardLimit;
        var status = utilization switch
        {
            >= 1.0 => "critical",
            >= QueueGuard.EmergencyRatio => "warning",
            >= QueueGuard.WarnRatio => "elevated",
            _ => "healthy",
        };
        return new QueueHealth(depth, hardLimit, utilization, status);
    }
}
